compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2358,13 +2358,20 @@ impl<'a> Parser<'a> {
     /// Parses a `let $pat = $expr` pseudo-expression.
     /// The `let` token has already been eaten.
     fn parse_let_expr(&mut self, attrs: AttrVec) -> PResult<'a, P<Expr>> {
-        if !self.let_expr_allowed {
-            self.struct_span_err(
-                self.prev_token.span,
-                "expected expression, found `let` statement",
-            )
-            .emit();
+        // Catches things like `if let Some(_) = _opt && [1, 2, 3][let _ = ()] = 1`
+        let is_in_a_let_chains_context_but_nested_in_other_expr = self.let_expr_allowed
+            && !matches!(
+                self.prev_token.kind,
+                TokenKind::AndAnd
+                    | TokenKind::CloseDelim(Delimiter::Brace)
```

Comment:
Out of curiosity, what situation does the `}` case protect from?

Reply:
Blocks in the right-hand-side followed be another `let` check.

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=806f4c7469764edaeb37224cf6657154

Reply:
I'm having trouble understanding this, can you clarify? 

From what I can tell, it is looking at the token just before the let.

From the example given:

```rust
fn main() {
    let opt = Some(1i32);
    if let Some(n) = opt
        && let add = {
            dbg!(n);
            |local_n| local_n + 1
        }
        && let new_n = add(1)
        && new_n == 2
    {
        println!("Peek a boo");
    }
}
```

In these let chains, the only tokens that appear before `let` is `if` and `&&`.  It's not clear to me when the close brace can precede a `let`.

Reply:
Without this change, CI didn't stop issuing errors like in https://github.com/rust-lang/rust/runs/7098719231?check_suite_focus=true.

Couldn't reproduce the behaviour locally

Reply:
OK, I'm not sure what the issue that was encountered, but I can't reproduce it with a full docker run. I opened #99731 to remove it just to avoid any possible confusion.